File: front/web/lib/operation.ts
/**
 * The Operation DSL — the Workbench /operation projection source (AIDOS step S10).
 *
 * WORKFLOW-AS-ARTIFACT (KRD §24.3, §93): an operation body is an ordered list of
 * TYPED step verbs — validate / authorize / read / mutate / branch / return —
 * stored content-addressed in kernel.operation, never free code. Its truth is the
 * fixture `state → command → events` (N2). This module holds the DECLARED
 * description of that shape (the six verbs — field-for-field with
 * back/kernel/operation), the §93 createOrder anchor, and a tiny deterministic
 * interpreter that mirrors Go's operation.Interpret over MOCK deps — so the
 * /operation panel renders the step pipeline and the createOrder/happy event trace
 * before the live kernel.operation SELECT wiring lands.
 *
 * DETERMINISM-FIRST (CLAUDE.md §6/§8): a static registry + a pure interpreter (no
 * clock, no rng, no I/O). The reproducibility mirror lib/operation.test.ts pins the
 * verb set, the anchor's pipeline, and the happy trace against the Go fixture
 * (events [OrderCreated, CartCleared], status "pending", total 15, authorize before
 * any mutate).
 *
 * READ-ONLY (CLAUDE.md §7 ui-completeness): /operation is a visualization; it
 * projects the kernel.operation AST and the fixture RESULT, it does not run truth
 * writes and exposes no capability — an operation AST is written only by the aidos
 * CLI through an approved ChangeSet, and a mutate reaches the world only through a
 * Mutator seam (below the waterline). There is no headless capability hidden here.
 */

/** The six operation verbs (matches operation.go stepKinds, canonical order). */
export const STEP_KINDS = [
	"validate",
	"authorize",
	"read",
	"mutate",
	"branch",
	"return",
] as const;
export type StepKind = (typeof STEP_KINDS)[number];

/** A typed step node — a discriminated union, the closed grammar. */
export type Step =
	| { kind: "validate"; schema: string }
	| { kind: "authorize"; policy: string }
	| { kind: "read"; entity: string; where: Record<string, unknown>; as: string }
	| {
			kind: "mutate";
			entity: string;
			op: "create" | "clear";
			data?: Record<string, unknown>;
			where?: Record<string, unknown>;
			as?: string;
	  }
	| { kind: "return"; ref: string };

/** A typed Operation (matches the Go Operation struct). */
export interface Operation {
	name: string;
	input: string;
	steps: Step[];
	emits: string[];
}

/** Is k one of the six verbs? (mirrors operation.IsStepKind). */
export function isStepKind(k: string): k is StepKind {
	return (STEP_KINDS as readonly string[]).includes(k);
}

/**
 * The KRD §93 anchor operation createOrder, VERBATIM (no invented step/event):
 *   input CreateOrderInput; steps [validate, authorize, read, mutate, mutate,
 *   return]; emits [OrderCreated, CartCleared].
 */
export const CREATE_ORDER: Operation = {
	name: "createOrder",
	input: "CreateOrderInput",
	steps: [
		{ kind: "validate", schema: "CreateOrderInput" },
		{ kind: "authorize", policy: "canPlaceOrder" },
		{
			kind: "read",
			entity: "Cart",
			where: { id: "$.input.cartId" },
			as: "$.cart",
		},
		{
			kind: "mutate",
			entity: "Order",
			op: "create",
			data: {
				userId: "$.auth.user.id",
				items: "$.cart.items",
				status: "pending",
				// total = sum($.cart.items, "price") — computed by the mutate seam.
			},
			as: "$.order",
		},
		{ kind: "mutate", entity: "Cart", op: "clear", where: { id: "$.cart.id" } },
		{ kind: "return", ref: "$.order" },
	],
	emits: ["OrderCreated", "CartCleared"],
};

/** One ordered pipeline chip for the panel: the verb + its one-line target. */
export interface PipelineChip {
	kind: StepKind;
	label: string;
}

/** The ordered pipeline chips of an operation (the visual step list). */
export function pipeline(op: Operation): PipelineChip[] {
	return op.steps.map((s): PipelineChip => {
		switch (s.kind) {
			case "validate":
				return { kind: "validate", label: s.schema };
			case "authorize":
				return { kind: "authorize", label: s.policy };
			case "read":
				return { kind: "read", label: `${s.entity} → ${s.as}` };
			case "mutate":
				return { kind: "mutate", label: `${s.entity} ${s.op}` };
			case "return":
				return { kind: "return", label: s.ref };
			default:
				// The verb set is closed; an unknown kind is a typed failure (mirrors
				// the Go step table's explicit default → unknown step kind).
				throw new Error(
					`operation: unknown step kind ${(s as { kind: string }).kind}`,
				);
		}
	});
}

// ── A tiny deterministic interpreter mirroring Go's operation.Interpret ───────

/** The $-rooted state bag ($.input, $.auth, named slots). */
type StateRoot = Record<string, unknown>;

function resolve(path: string, root: StateRoot): unknown {
	if (!path.startsWith("$")) return undefined;
	const segs = path.split(".").slice(1);
	let cur: unknown = root;
	for (const s of segs) {
		if (s === "length") {
			if (Array.isArray(cur) || typeof cur === "string") return cur.length;
			return undefined;
		}
		if (typeof cur !== "object" || cur === null) return undefined;
		cur = (cur as Record<string, unknown>)[s];
	}
	return cur;
}

function resolveMap(
	m: Record<string, unknown> | undefined,
	root: StateRoot,
): Record<string, unknown> {
	const out: Record<string, unknown> = {};
	if (!m) return out;
	for (const [k, v] of Object.entries(m)) {
		out[k] = typeof v === "string" && v.startsWith("$") ? resolve(v, root) : v;
	}
	return out;
}

/** The mock deps the fixture passes (mirrors the Go mockDeps). */
export interface MockDeps {
	authorizeAllow: boolean;
	cart: unknown;
	calls: string[];
}

/** The result of an interpreted operation (mirrors the Go Result). */
export interface RunResult {
	events: string[];
	ref: Record<string, unknown> | null;
	denied: boolean;
	calls: string[];
}

/**
 * run walks an operation's steps in order over the mock deps, returning the ordered
 * events and the return ref — the deterministic mirror of Go's Interpret. An
 * authorize DENY short-circuits (events stay empty, no mutate runs).
 */
export function run(
	op: Operation,
	state: StateRoot,
	deps: MockDeps,
): RunResult {
	const root: StateRoot = { ...state };
	const events: string[] = [];
	let ref: Record<string, unknown> | null = null;

	for (const step of op.steps) {
		switch (step.kind) {
			case "validate":
				deps.calls.push("validate");
				break;
			case "authorize":
				deps.calls.push("authorize");
				if (!deps.authorizeAllow) {
					return { events: [], ref: null, denied: true, calls: deps.calls };
				}
				break;
			case "read": {
				deps.calls.push("read");
				if (step.as) root[step.as.slice(2)] = deps.cart;
				break;
			}
			case "mutate": {
				deps.calls.push("mutate");
				const data = resolveMap(step.data, root);
				if (step.op === "create") {
					let total = 0;
					const items = data.items;
					if (Array.isArray(items)) {
						for (const it of items) {
							if (it && typeof it === "object" && "price" in it) {
								const p = (it as { price: unknown }).price;
								if (typeof p === "number") total += p;
							}
						}
					}
					const created = { status: data.status, total };
					events.push("OrderCreated");
					if (step.as) root[step.as.slice(2)] = created;
				} else {
					events.push("CartCleared");
				}
				break;
			}
			case "return": {
				const v = resolve(step.ref, root);
				ref =
					v && typeof v === "object" ? (v as Record<string, unknown>) : null;
				break;
			}
		}
	}
	return { events, ref, denied: false, calls: deps.calls };
}

/** The createOrder/happy given state: authed u1, cart c1 with two priced items. */
export const HAPPY_STATE: StateRoot = {
	input: { cartId: "c1" },
	auth: { user: { id: "u1" } },
};

/** The cart the mock Reader yields for createOrder/happy. */
export const HAPPY_CART = {
	id: "c1",
	userId: "u1",
	items: [{ price: 10 }, { price: 5 }],
};

/** The fixture trace the /operation panel renders: the createOrder/happy run + its PASS verdict. */
export interface FixtureTrace {
	id: string;
	events: string[];
	status: unknown;
	total: unknown;
	authorizeBeforeMutate: boolean;
	pass: boolean;
}

/**
 * createOrderHappyTrace runs the createOrder/happy fixture through the mirror
 * interpreter and computes the PASS verdict exactly as the Go fixture asserts:
 * events == [OrderCreated, CartCleared], status "pending", total 15, authorize ran
 * before any mutate. Deterministic — same inputs, same trace.
 */
export function createOrderHappyTrace(): FixtureTrace {
	const deps: MockDeps = {
		authorizeAllow: true,
		cart: HAPPY_CART,
		calls: [],
	};
	const res = run(CREATE_ORDER, HAPPY_STATE, deps);
	const ai = res.calls.indexOf("authorize");
	const mi = res.calls.indexOf("mutate");
	const authorizeBeforeMutate = ai >= 0 && mi >= 0 && ai < mi;
	const status = res.ref?.status;
	const total = res.ref?.total;
	const pass =
		res.events.length === 2 &&
		res.events[0] === "OrderCreated" &&
		res.events[1] === "CartCleared" &&
		status === "pending" &&
		total === 15 &&
		authorizeBeforeMutate;
	return {
		id: "createOrder/happy",
		events: res.events,
		status,
		total,
		authorizeBeforeMutate,
		pass,
	};
}

File: back/kernel/operation/operation.go
// Package operation is the AIDOS Operation DSL interpreter (KRD §24.3, §93): the
// N2 (Workflow) executor. An operation body is an ordered list of typed steps —
// validate / authorize / read / mutate / branch / return — read from its Kernel
// AST; this package walks them deterministically as `state → command → events`.
//
// THE SHAPE (KRD §93). An Operation is {Name, Input, Steps[], Emits[]}:
//   - Input  — the named input schema (e.g. "CreateOrderInput");
//   - Steps  — the ordered verbs the interpreter walks, each a typed Step node;
//   - Emits  — the declared event names the operation may emit (a contract on the
//     event surface; the interpreter's emitted events are a subsequence of Emits).
//
// THE SIX VERBS (KRD §24.3). validate (shape-check the input) · authorize
// (delegate to the Policy DSL — DENY short-circuits the pipeline) · read (load an
// entity into a named state slot) · mutate (create/clear an entity, emitting an
// event) · branch (conditional — supported in the table, exercised only when a
// fixture needs it; no invented semantics) · return (name the result ref).
//
// STATE → COMMAND → EVENTS (the N2 truth form). Interpret threads a State bag
// ($.input, $.auth, named slots like $.cart / $.order) through the steps and
// returns the ORDERED event list plus the return Result. Deterministic: the emit
// order mirrors the step order; same (op, state, deps) ⇒ same (events, result).
//
// THE SEAMS (mocked this step). Side-effecting verbs reach the world ONLY through
// injected deps interfaces (Validator/Authorizer/Reader/Mutator). The package is
// PURE — no DB, no HTTP, no clock, no RNG (CLAUDE.md §6 determinism-first). The
// real Policy ∀ evaluation, the real Expr `sum`, the real sqlc read/write are
// later teeth; here the fixture passes mocks at each seam.
//
// THE WALL (CLAUDE.md §2). The interpreter writes no truth: a mutate is a call on
// the injected Mutator (a projection seam, below the waterline), never a direct
// write to a kernel/mirrors/fitness schema. The operation AST it reads is a SOURCE
// row written only by the aidos CLI through an approved ChangeSet.
//
// REUSE, DON'T REINVENT (ADR 0007). The interpreter is table-driven (a kind →
// evaluator dispatch table, stepTable in table.go), not a switch scattered across
// files — verbs are data. The $-rooted selector semantics match back/kernel/expr
// and back/kernel/policy (one selector language across the DSLs). No new ADR: this
// freezes a new artifact (the Operation interpreter); it shifts no prior contract.
package operation

import "errors"

// ErrAuthorizationDenied is the typed error the authorize verb returns when the
// injected Authorizer denies. It short-circuits the pipeline: no later step runs,
// no event is emitted, no mutate is attempted (the §93 "any DENY blocks" law at
// the operation boundary). Callers test with errors.Is(err, ErrAuthorizationDenied).
var ErrAuthorizationDenied = errors.New("operation: authorization denied")

// ErrUnknownStepKind is the typed error the step table returns for an unrecognised
// verb — a kind with no evaluator. It is a typed FAILURE, never a silent skip:
// the table's explicit default. (A well-formed AST never triggers it; it guards
// against a corrupt/unknown step kind decoded from JSONB.)
var ErrUnknownStepKind = errors.New("operation: unknown step kind")

// MutateOp is the closed set of mutate operations (KRD §93: create, clear).
type MutateOp string

const (
	// MutateCreate creates an entity row (the Order create → OrderCreated).
	MutateCreate MutateOp = "create"
	// MutateClear clears an entity (the Cart clear → CartCleared).
	MutateClear MutateOp = "clear"
)

// Operation is one content-addressed workflow body (KRD §24.3, §93).
// {Name, Input, Steps, Emits}. It is read from the kernel.operation AST; this
// package walks Steps in order.
type Operation struct {
	Name  string
	Input string
	Steps []Step
	Emits []string
}

// Result is the return of an interpreted operation: the resolved return ref (the
// named slot the `return` verb points at, e.g. $.order), surfaced as a map for the
// caller to read (result.Ref["status"], result.Ref["total"]). RefName records
// which slot was returned (for diagnostics / the Workbench trace).
type Result struct {
	RefName string
	Ref     map[string]any
}
